### ISSUE TITLE
Proposal: Simplify Grid Rows, Columns Definition and Implementation

### ISSUE THREAD

**Issue by rizamarhaban** [microsoft/microsoft-ui-xaml#2094]
# Proposal: Simplify Grid Rows, Columns Definition and Implementation 

## Summary
This proposal is to add feature to simplify the syntax of `Grid` Rows, Columns and its implementation:

- Define `Grid` Rows and Columns definitions inline. Can use comma or space as the delimeter.
- Allowing to add _variable_ for each row and column index definition value. I propose to use a colon simply `[variable name]:[value]`. See example below on this.
- Adding new attribute `Grid.RowColumn` besides `Grid.Row` and `Grid.Column`.
- Using the variable as well as index number to assign rows and columns in implementation.

## Rationale

* Currently it is difficult to refactor Grid rows and column. Imagine having so many rows and columns and we want to insert a row or column, we usually do this:
   - Add `RowDefinition` and `ColumnDefinition`.
   - Reassign the indexes of the implementation.
* Switching index is difficult and require to reassign the indexes and we usually make mistake because of the index number.
* This will add productivity by easy to refactor and easy to trace rows and columns.

## Scope

| Capability | Priority |
| :---------- | :------- |
| Original syntax not changed | Must |
| Add this feature once to simplify and easy refactoring of Grid| Should |
| This model can be used for other similar cases if any | Could |

### Define Grid Rows and Columns Inline
```xaml
<Grid ColumnDefinitions="a:1* b:2* c:Auto d:* e:300"
      RowDefinitions="a:1* Auto i:25 whatever:14 k:20">
</Grid>
```

### Grid Row and Column Implementation
When we need to reference on which rows and column, we just do:

```xaml
<TextBlock Grid.Row="a" Grid.Column="b" />
<TextBlock Grid.Row="whatever" Grid.Column="e" />
<TextBlock Grid.Row="1" Grid.Column="c" />  <!-- using index number on Row -->
```

Or if no variable, simply fallback to use index number.

### New Attribute `Grid.RowColumn`
This to simplify without the assignment for the row and column:

```xaml
<TextBlock Grid.RowColumn="a b" />
<TextBlock Grid.RowColumn="whatever e" />
<TextBlock Grid.RowColumn="1 c" />
```

With these hopefully easy to refactor as well.

**Comment by japf** [comment]
This looks like a duplicate of #673

**Comment by Felix-Dev** [comment]
@japf It's not. Issue #673 is about simplifying the syntax used to define rows and columns.
> Ability for rows and columns to be defined as a collection or list of values (using CreateFromString attribute)
Adding a name to columm so we can use the name to reference a row/column instead of its index in the row/column collection is purposefully *not* part of that proposal. This issue here proposes adding that naming capability to rows/columns so that changes like inserting new rows/columns in the grid don't require changing the indices in affected `<Control Grid.Row/Column=.../>` XAML.

The example you see in this proposal shows how naming rows/columns could work alongside the new syntax introduced in #673.

**Comment by mdtauk** [comment]
The name should change to reflect to focus on the unique proposal here.  Naming Grid Rows and Columns for Simplified Syntax.

**Comment by Felix-Dev** [comment]
Yes, the name of the proposal should be changed. An example name could be: "Able to refer to rows / columns by name" (based on the Make Grid Better proposal).

**Comment by ranjeshj** [comment]
It does look very similar to #673 except for the naming aspect. I like the @mdtauk 's idea of scoping this down to just the naming and keep the rest of the discussions in the other proposal. Thanks.

**Comment by Felix-Dev** [comment]
Given the rationale of this porposal it has to be changed then to include examples/API how grid rows/columns can be assigned a name with the current grid API/syntax. Since issue #673 has been greenlighted based on my understanding the idea
> Allowing to add variable for each row and column index definition value. I propose to use a colon simply [variable name]:[value]

should probably also be kept in this proposal

**Comment by ranjeshj** [comment]
@rizamarhaban Thank you for filing the proposal. Can you please update the proposal and scope it down to just naming rows and columns (on top of what is proposed in #673 ) ?

**Comment by rizamarhaban** [comment]
@Felix-Dev @ranjeshj I include the syntax simplification like similar in #673 because has not been implemented and I have the same idea but with adding variables or naming row/column. I also add the new attribute Grid.RowColumn which is not the same with #673. I can update the title to reflect this, but this is not just simply inlining the row/col values.

**Comment by Felix-Dev** [comment]
@rizamarhaban You seem to propose two new features here - naming rows/columns and a new API for a compact row/column specification (Grid.RowColumn). The naming feature alone should be in its own proposal as suggested by @ranjeshj as it raises a couple of questions worth thinking about. (I.e. will it only work for the new compact syntax or also for the current grid syntax? If the latter, do we need to add new API and how would that look like?)

Given this, you should probably split this issue into two issues...one issue for the naming feature and and one issue for the new Grid.RowColumn API.

**Comment by rizamarhaban** [comment]
@Felix-Dev yes it's 2 new features and will add into the scope.

**Comment by michael-hawker** [comment]
Should also be an option name:

```xml
<Grid ColumnDefinitions="1* b:2* Auto d:* 300"
      RowDefinitions="1* Auto i:25 14 20">
</Grid>
```

**Comment by rizamarhaban** [comment]
@michael-hawker yes it is. In the example, if there is no variable name then fallback to indexes. We can use both indexes or the name if assigned.

**Comment by Felix-Dev** [comment]
@rizamarhaban Please answer the following questions:
Will your naming proposal only work for the new compact syntax or also for the current grid syntax? If the latter, do we need to add new API and how would that look like?

You are proposing a feature wich would make the current Grid syntax and the new compact syntax in issue #673 no longer equivalent. In other words, the way I'm reading your naming proposal currently, you are a adding a feature available to the new compaxt syntax *only*. Issue #673 only proposes a sytax simplication, it does not add new features exclusive to the new compact syntax. Your current proposal would change that and thus this should be discussed.

**Comment by rizamarhaban** [comment]
@Felix-Dev 

> Will your naming proposal only work for the new compact syntax or also for the current grid syntax? If the latter, do we need to add new API and how would that look like?

Technically, this can be use for the current syntax as well. Alternatively, both new compact syntax and current syntax may implement this. Meaning, there will be new API requirements. As this won't break any changes of the current XAML, need to discuss what the API may look like. I will update the Example to include the current syntax as well.

**Comment by Felix-Dev** [comment]
@rizamarhaban Thanks for updating 🙂

**Comment by jtorjo** [comment]
This looks like a very very good idea. This way, when re-arranging rows/columns, it will be very easy.

One thing that comes to mind is adding a grid splitter (which takes one row or column) <- this right now requires me manually editing the xaml and hope I don't miss anything. With this proposal, I would not need any xaml editing.

Also, very useful when you need to quickly re-arrange rows (you realize some low row needs to come to the top or so).

**Comment by anawishnoff** [comment]
Thanks for filing this, @rizamarhaban! Sorry I am just getting around to seeing it. The naming idea (as others mentioned) was brought up in the initial Make Grid Better proposal #54, and it seems like a popular want to help deal with large numbers of rows and columns. I like that your proposal accounts for making sure that it works with the old syntax and the new, compact syntax - we don't want to cause breaking changes or leave pure UWP (non-WinUI 3) apps out of the feature.

In terms of the RowColumn attribute, I was wondering what the motivation was. Is it just a shorter way of assigning placement to elements, instead of writing Grid.Column="0" and Grid.Row="0"? To me, this seems like it would take almost the same amount of effort to write and I'm not sold that it's something that needs to be implemented... feel free to prove me wrong! Would love to hear more arguments for this.

One more detail question about your proposed changes: I'm wondering what happens behind the scenes when a name for a row is omitted, as shown here with `Auto`:
```
<Grid ColumnDefinitions="a:1* b:2* c:Auto d:* e:300"
      RowDefinitions="a:1* Auto i:25 whatever:14 k:20">
</Grid>
```

You use the index number to refer to it here:
```<TextBlock Grid.Row="1" Grid.Column="c" />  <!-- using index number on Row -->```

Does this mean that even if a row/column _does_ have a name, you can still use its index to refer to it? I assume that the Name property for a RowDefinition or ColumnDefinition is a string. Would the index get converted into a string and used for the Name property, or would it just be the general rule that rows/columns can be referred to by their index or Name? If so, that leads to more questions about accessing rows and columns in the code-behind... 

Maybe I'm thinking too far ahead, but let me know if anyone has thoughts on these points :)

**Comment by michael-hawker** [comment]
@anawishnoff I could see the Row/Column attached properties still being defined as integers, it'd just be the XAML [TypeConverter](https://docs.microsoft.com/en-us/uwp/api/windows.foundation.metadata.createfromstringattribute) that knows how to do the magic with the string to look-up an index. Then there'd be a method like `GetIndexFromName(string)` on Grid that's used in that function as well as for code-behind use?

**Comment by rizamarhaban** [comment]
@anawishnoff the motivation is:
1. Easy refactoring (finding row and column). We used to marked it with `<!-- ... -->` for rows or columns for easy searching. Especially on a large Grid.
2. Suppose we wanted to insert a row or column on existing grid, we don't have to reassigned each index, however that is what happening currently. With this proposal, we can just add new row or column with variable naming and use the name for Grid.Row/Column on other elements. This will not disturb the other rows/columns. Or `Grid.RowColumn` for short.
3. The fallback to indexes is just a lookup as mentioned by @michael-hawker. The important thing is to look at the implementation of the naming on other elements, and not from the definitions only. Because defining rows/columns still uses index behind the scene. Doesn't matter if we reorder the definitions, the index is always sequential. So if we use naming on the definition, then reordering the definitions will also effect the implementation. Currently it is not, unless we specify Height or Width, then we can see the difference. Example;

```xml
<Grid>
    <Grid.RowDefinitions>
        <RowDefinition  />
        <RowDefinition Height="Auto" />
        <RowDefinition Height="25" />
        <RowDefinition  />
    </Grid.ColumnDefinitions>
</Grid>
```
In the above example, if we reorder index 0 and 3, it doesn't have any changes on the UI. However, index 2 and 2 might change the UI because height different. However, if we can do like this:

```xml
<Grid>
    <Grid.RowDefinitions>
        <RowDefinition x:Name="a"  />
        <RowDefinition Height="Auto" />
        <RowDefinition Height="25" />
        <RowDefinition x:Name="b" />
    </Grid.ColumnDefinitions>
</Grid>
```
and I switch the definitions to be like this:

```xml
<Grid>
    <Grid.RowDefinitions>
        <RowDefinition x:Name="b"  />
        <RowDefinition Height="Auto" />
        <RowDefinition Height="25" />
        <RowDefinition x:Name="a" />
    </Grid.ColumnDefinitions>
</Grid>
```
Then all the UI which uses Grid.Row="a" or Grid.Row="b" will switch as well. Nice right. No need to replace the index number from 3 to 0 and 0 to 3. Unless they did not define naming like on row index 1 and 2. That will need to follow the current existing rule on Grid index. Also, if we need to do like this:

```xml
<Grid>
    <Grid.RowDefinitions>
        <RowDefinition x:Name="a"  />
        <RowDefinition Height="Auto" />
        <RowDefinition x:Name="c" Height="Auto"  />
        <RowDefinition Height="25" />
        <RowDefinition x:Name="b" />
    </Grid.ColumnDefinitions>
</Grid>
```
There will be no changes on Grid.Row 'a' and 'b' elements. However, we need to replace the index of row '3' to index '4' to adjust the Grid.Row element. :) That's the motivation. Let's make it simple like this, e.g.:

```xml
<!--     Index:            0            1      2 -->
<Grid RowDefinitions="awesome:Auto lovely:Auto *">

<TextBlock Grid.RowColumn="awesome,0" Text="Awesome..." />
<TextBlock Grid.RowColumn="lovely,0" Text="Lovely..." />
<StackPanel Grid.RowColumn="2,0" />
</Grid>
```
This will show Awesome and Lovely on each row. And someday we can just switch the definition to be like this:

```xml
<!--     Index:            0            1      2 -->
<Grid RowDefinitions="lovely:Auto awesome:Auto *">

<TextBlock Grid.RowColumn="awesome,0" Text="Awesome..." />
<TextBlock Grid.RowColumn="lovely,0" Text="Lovely..." />
<StackPanel Grid.RowColumn="2,0" />
</Grid>
```
The UI changes to show Lovely and Awesome on each row. And there you go.

4. And possibly more usage in the future.

**Comment by anawishnoff** [comment]
@michael-hawker @rizamarhaban 

I definitely think it would be necessary to have a GetIndexFromName() method here, which makes things a bit easier. Thank you for all of the examples - it definitely clears up how helpful named rows and columns would be in terms of switching around the Grid and interacting with it.

The unnamed row/column scenario is a bit messy, but it would be the developer's prerogative as to whether or not they want to name their rows/columns for more ease.

I'm gonna push just a little bit harder on the Grid.RowColumn attribute 😊: I do get what you're saying that it may make refactoring easier for large grids, although for searching within a grid I feel like searching for individual named rows and columns would be similarly easy. Is that enough of a reason to implement a whole other attribute into Grid which already has such a large API? Do you know of any other Xaml controls which implement similar attributes? Would need some more convincing on this one.

**Comment by michael-hawker** [comment]
Yeah, I don't see as much of a need for `RowColumn`, especially as I'm sure @mrlacey could come up with some amazing helper as part of his extensions for moving items around in a Grid for refactoring... 😋

**Comment by rizamarhaban** [comment]
@anawishnoff @michael-hawker yeah, fair enough.

**Comment by mrlacey** [comment]
The `RowColumn` idea seems like it would be easy to try out with a custom attached property. It _feels_ like the kind of thing that could/should be tried out before a lot of time is spent on incorporating it into something larger (like WinUI.) 
After trying it out, it would be good to get feedback on what it's like to use, how beneficial it is, whether new developers (unfamiliar with the syntax) find it easy to understand or confusing because it's too different to what they already know, etc.

**Comment by murilocurti** [comment]
A good point is that inline syntax will enable to configure the grid using Style Setters:

<Setter Property="ColumnDefinitions" Value="*,*,*" />

**Comment by sjb-sjb** [comment]
Suggesting that the syntax be expanded to include a trailing &gt;&eq;_n_ and/or &lt;&eq;_c_ representing MinHeight and MaxHeight for each row/column specifier. I have done this successfully in my own version of this feature. 

Another suggestion, if it is not too much of a distraction, is to allow just A as a short form of Auto.
